### PR TITLE
object.d: Improve coverage by adding some missing unittests

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

### CODE REVIEW

**Comment by 12345swordy** [comment]
cc @RazvanN7

File: src/object.d [hunk 1]
```diff
@@ -266,13 +306,40 @@ bool opEquals(const Object lhs, const Object rhs)
...
 
 void setSameMutex(shared Object ownee, shared Object owner)
 {
     _d_setSameMutex(ownee, owner);
 }
 
+@system unittest
+{
+    shared Object obj1 = new Object;
+    synchronized class C {
```
Comment (RazvanN7) [inline src/object.d:329]: Bracket on a new line.

File: src/object.d [hunk 2]
```diff
@@ -266,13 +306,40 @@ bool opEquals(const Object lhs, const Object rhs)
...
+        void bar() { }
```
Comment (RazvanN7) [inline src/object.d:330]: no space between brackets.

File: src/object.d [hunk 3]
```diff
@@ -383,6 +465,35 @@ class TypeInfo
         }
     }
 
+    @system unittest
+    {
+        class _TypeInfo_Dummy : TypeInfo {
```
Comment (RazvanN7) [inline src/object.d:470]: Bracket on new line.

File: src/object.d [hunk 4]
```diff
@@ -424,6 +535,80 @@ class TypeInfo
     @property immutable(void)* rtInfo() nothrow pure const @safe @nogc { return rtinfoHasPointers; } // better safe than sorry
 }
 
+@system unittest
+{
+    class _TypeInfo_Dummy : TypeInfo {
```
Comment (RazvanN7) [inline src/object.d:540]: Bracket on new line.

File: src/object.d [hunk 5]
```diff
@@ -424,6 +535,80 @@ class TypeInfo
...
+    auto dummy = new _TypeInfo_Dummy();
+    cast(void)dummy.initializer(); // For coverage completeness
+
+    assert(dummy.rtInfo() is rtinfoHasPointers);
+    assert(typeid(void).rtInfo() is rtinfoNoPointers);
+
+    assert(dummy.tsize() == 0);
+
+    bool gotCaught;
+    try {
```
Comment (RazvanN7) [inline src/object.d:552]: Bracket on new line.

File: src/object.d [hunk 6]
```diff
@@ -424,6 +535,80 @@ class TypeInfo
...
+        assert(typeid(void) is ti1);
+
+        assert(ti1 !is null);
+        assert(ti2 is null);
+    }
+}
+
+@system unittest
+{
+    class _ZypeInfo_Dummy : TypeInfo {
```
Comment (RazvanN7) [inline src/object.d:585]: Bracket on new line.

File: src/object.d [hunk 7]
```diff
@@ -424,6 +535,80 @@ class TypeInfo
...
+    assert(typeid(void) > dummy2);
+    assert(dummy2 < typeid(void));
+}
+
+@safe unittest
+{
+    enum unittest_sym_name = __traits(identifier, __traits(parent, (){}));
+    enum fqn_unittest = "object." ~ unittest_sym_name; // object.__unittest_LX_CY
+
+    class _TypeInfo_Dummy : TypeInfo {
```
Comment (RazvanN7) [inline src/object.d:600]: Bracket on new line.

File: src/object.d [hunk 8]
```diff
@@ -437,15 +622,114 @@ class TypeInfo_Enum : TypeInfo
...
+        E e2 = E.B;
+
+        typeid(E).swap(&e1, &e2);
+        assert(e1 == E.B);
+        assert(e2 == E.A);
+    }
+
     override @property inout(TypeInfo) next() nothrow pure inout { return base.next; }
+
+    @system unittest
```
Comment (RazvanN7) [inline src/object.d:708]: Why can't this be `@safe`?

File: src/object.d [hunk 9]
```diff
@@ -437,15 +622,114 @@ class TypeInfo_Enum : TypeInfo
...
+    @safe unittest
+    {
+        enum E { A, B, C }
+
+        assert(typeid(E).flags == 0);
+    }
+
     override const(OffsetTypeInfo)[] offTi() const { return base.offTi; }
+
+    @system unittest
```
Comment (RazvanN7) [inline src/object.d:726]: Why can't this be `@safe`?

File: src/object.d [hunk 10]
```diff
@@ -1295,6 +1637,18 @@ class TypeInfo_Interface : TypeInfo
     }
 }
 
+@safe unittest
```
Comment (RazvanN7) [inline src/object.d:1640]: If this is `@safe`, what prevents other unittests that call `toString`/`typeid` to also be `@safe`?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Thanks you for this! Please mind the style changes and also double check if other `@system` unittests cannot be `@safe`.

File: src/object.d (hunk 10 above)
Comment (ljmf00) [reply src/object.d:1646]: Yeah, here is actually possible to call toString under @safe, for example for Enum `typeid()`. Changed that unittest. Although as I stated above, I'll later change TypeInfo to be @safe on some methods that currently are not.

File: src/object.d (hunk 1 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 2 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 3 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 4 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 5 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 6 above)
Comment (ljmf00) [reply src/object.d]: Done

File: src/object.d (hunk 7 above)
Comment (ljmf00) [reply src/object.d]: Done

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**